File: app/src/main/java/com/espressif/ui/activities/ProofOfPossessionActivity.java
package com.espressif.ui.activities;

import android.content.Intent;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.widget.ContentLoadingProgressBar;

import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.espressif.AppConstants;
import com.espressif.provision.Provision;
import com.espressif.provision.R;
import com.espressif.provision.security.Security;
import com.espressif.provision.security.Security0;
import com.espressif.provision.security.Security1;
import com.espressif.provision.session.Session;
import com.espressif.provision.transport.ResponseListener;
import com.google.protobuf.InvalidProtocolBufferException;

import avs.Avsconfig;

import static com.espressif.avs.ConfigureAVS.AVS_CONFIG_PATH;

public class ProofOfPossessionActivity extends AppCompatActivity {

    private static final String TAG = "Espressif::" + ProofOfPossessionActivity.class.getSimpleName();

    private Button btnNext;
    private TextView textDeviceName;
    private EditText etDeviceKey;
    private ContentLoadingProgressBar progressBar;

    private String securityVersion, transportVersion;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_pop);

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle(R.string.title_activity_pop);
        setSupportActionBar(toolbar);

        btnNext = findViewById(R.id.btn_next);
        textDeviceName = findViewById(R.id.device_name);
        etDeviceKey = findViewById(R.id.et_pop);
        progressBar = findViewById(R.id.progress_indicator);

        String deviceName = getIntent().getStringExtra(AppConstants.KEY_DEVICE_NAME);
        transportVersion = getIntent().getStringExtra(Provision.CONFIG_TRANSPORT_KEY);
        securityVersion = getIntent().getStringExtra(Provision.CONFIG_SECURITY_KEY);

        progressBar.setVisibility(View.INVISIBLE);
        textDeviceName.setText(deviceName);
        btnNext.setOnClickListener(nextBtnClickListener);
        String key = getString(R.string.proof_of_possesion);

        if (!TextUtils.isEmpty(key)) {

            etDeviceKey.setText(key);
            etDeviceKey.setSelection(etDeviceKey.getText().length());
        } else {
            btnNext.setEnabled(false);
            btnNext.setAlpha(0.5f);
        }

        etDeviceKey.addTextChangedListener(new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {

                String pop = s.toString();

                if (TextUtils.isEmpty(pop)) {

                    btnNext.setEnabled(false);
                    btnNext.setAlpha(0.5f);

                } else {

                    btnNext.setEnabled(true);
                    btnNext.setAlpha(1f);
                }
            }
        });
    }

    @Override
    public void onBackPressed() {
        BLEProvisionLanding.isBleWorkDone = true;
        super.onBackPressed();
    }

    private View.OnClickListener nextBtnClickListener = new View.OnClickListener() {

        @Override
        public void onClick(View v) {

            btnNext.setEnabled(false);
            btnNext.setAlpha(0.5f);
            progressBar.setVisibility(View.VISIBLE);

            final String pop = etDeviceKey.getText().toString();
            Log.d(TAG, "POP : " + pop);
            boolean shouldCreateSession = false;

            if (transportVersion.equals(Provision.CONFIG_TRANSPORT_BLE)) {

                if (BLEProvisionLanding.security == null) {

                    if (securityVersion.equals(Provision.CONFIG_SECURITY_SECURITY1)) {
                        BLEProvisionLanding.security = new Security1(pop);
                    } else {
                        BLEProvisionLanding.security = new Security0();
                    }

                    shouldCreateSession = true;

                } else {

                    if (BLEProvisionLanding.session != null) {
                        getStatus();
                    } else {
                        shouldCreateSession = true;
                    }
                }

            } else {

                if (securityVersion.equals(Provision.CONFIG_SECURITY_SECURITY1)) {
                    Security security = new Security1(pop);
                } else {
                    Security security = new Security0();
                }
                // TODO Create session for SoftAP Transport
                Log.e(TAG, "Currently BLE transport is supported in AVS app.");
            }

            if (shouldCreateSession) {

                BLEProvisionLanding.session = new Session(BLEProvisionLanding.bleTransport, BLEProvisionLanding.security);
                BLEProvisionLanding.session.sessionListener = new Session.SessionListener() {

                    @Override
                    public void OnSessionEstablished() {
                        Log.d(TAG, "Session established");
                        getStatus();
                    }

                    @Override
                    public void OnSessionEstablishFailed(Exception e) {
                        Log.d(TAG, "Session failed");
                    }
                };
                BLEProvisionLanding.session.init(null);
            }
        }
    };

    private void getStatus() {

        Avsconfig.CmdSignInStatus configRequest = Avsconfig.CmdSignInStatus.newBuilder()
                .setDummy(123)
                .build();
        Avsconfig.AVSConfigMsgType msgType = Avsconfig.AVSConfigMsgType.TypeCmdSignInStatus;
        Avsconfig.AVSConfigPayload payload = Avsconfig.AVSConfigPayload.newBuilder()
                .setMsg(msgType)
                .setCmdSigninStatus(configRequest)
                .build();

        byte[] message = BLEProvisionLanding.security.encrypt(payload.toByteArray());

        BLEProvisionLanding.bleTransport.sendConfigData(AVS_CONFIG_PATH, message, new ResponseListener() {

            @Override
            public void onSuccess(byte[] returnData) {

                Avsconfig.AVSConfigStatus deviceStatus = processSignInStatusResponse(returnData);
                Log.d(TAG, "SignIn Status Received : " + deviceStatus);
                finish();

                if (deviceStatus.equals(Avsconfig.AVSConfigStatus.SignedIn)) {

                    goToProvisionActivity();

                } else {
                    goToLoginActivity();
                }
            }

            @Override
            public void onFailure(Exception e) {
                Log.d(TAG, "Error in getting status");
                e.printStackTrace();
            }
        });
    }

    private Avsconfig.AVSConfigStatus processSignInStatusResponse(byte[] responseData) {

        Avsconfig.AVSConfigStatus status = Avsconfig.AVSConfigStatus.InvalidState;
        byte[] decryptedData = BLEProvisionLanding.security.decrypt(responseData);

        try {

            Avsconfig.AVSConfigPayload payload = Avsconfig.AVSConfigPayload.parseFrom(decryptedData);
            Avsconfig.RespSignInStatus signInStatus = payload.getRespSigninStatus();
            status = signInStatus.getStatus();
            Log.d(TAG, "SignIn Status message " + status.getNumber() + status.toString());

        } catch (InvalidProtocolBufferException e) {
            e.printStackTrace();
        }
        return status;
    }

    private void goToLoginActivity() {

        Intent alexaProvisioningIntent = new Intent(getApplicationContext(), LoginWithAmazon.class);
        alexaProvisioningIntent.putExtras(getIntent());
        alexaProvisioningIntent.putExtra(AppConstants.KEY_PROOF_OF_POSSESSION, etDeviceKey.getText().toString());
        startActivity(alexaProvisioningIntent);
    }

    private void goToProvisionActivity() {

        Intent launchProvisionInstructions = new Intent(getApplicationContext(), ProvisionActivity.class);
        launchProvisionInstructions.putExtras(getIntent());
        launchProvisionInstructions.putExtra(AppConstants.KEY_PROOF_OF_POSSESSION, etDeviceKey.getText().toString());
        startActivityForResult(launchProvisionInstructions, Provision.REQUEST_PROVISIONING_CODE);
    }
}

File: app/src/main/java/com/espressif/ui/activities/ProofOfPossessionActivity.kt
// Copyright 2025 Espressif Systems (Shanghai) PTE LTD
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.espressif.ui.activities

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.content.res.AppCompatResources
import com.espressif.AppConstants
import com.espressif.provisioning.DeviceConnectionEvent
import com.espressif.provisioning.ESPConstants
import com.espressif.provisioning.ESPProvisionManager
import com.espressif.provisioning.listeners.ResponseListener
import com.espressif.wifi_provisioning.R
import com.espressif.wifi_provisioning.databinding.ActivityPopBinding
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

class ProofOfPossessionActivity : AppCompatActivity() {

    companion object {
        private val TAG: String = ProofOfPossessionActivity::class.java.simpleName
    }

    private lateinit var binding: ActivityPopBinding

    private var deviceName: String? = null
    private lateinit var provisionManager: ESPProvisionManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityPopBinding.inflate(layoutInflater)
        setContentView(binding.root)

        provisionManager = ESPProvisionManager.getInstance(applicationContext)
        deviceName = provisionManager.espDevice!!.deviceName
        initViews()
        EventBus.getDefault().register(this)

        if (!TextUtils.isEmpty(deviceName)) {
            val popText = getString(R.string.pop_instruction) + " " + deviceName
            binding.tvPop.text = popText
        }

        val pop = resources.getString(R.string.proof_of_possesion)

        if (!TextUtils.isEmpty(pop)) {
            binding.etPop.setText(pop)
            binding.etPop.setSelection(binding.etPop.text.length)
        }
        binding.etPop.requestFocus()
    }

    override fun onDestroy() {
        EventBus.getDefault().unregister(this)
        super.onDestroy()
    }

    override fun onBackPressed() {
        provisionManager.espDevice?.disconnectDevice()
        super.onBackPressed()
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: DeviceConnectionEvent) {
        Log.d(TAG, "On Device Connection Event RECEIVED : " + event.eventType)

        when (event.eventType) {
            ESPConstants.EVENT_DEVICE_DISCONNECTED -> if (!isFinishing) {
                showAlertForDeviceDisconnected()
            }
        }
    }

    private val nextBtnClickListener = View.OnClickListener {
        val pop = binding.etPop.text.toString()
        Log.d(TAG, "POP : $pop")

        binding.btnNext.layoutBtn.isEnabled = false
        binding.btnNext.layoutBtn.alpha = 0.5f
        binding.btnNext.progressIndicator.visibility = View.VISIBLE
        binding.tvErrorPop.visibility = View.INVISIBLE
        provisionManager.espDevice.proofOfPossession = pop
        provisionManager.espDevice.initSession(object : ResponseListener {

            override fun onSuccess(returnData: ByteArray?) {
                runOnUiThread {
                    binding.btnNext.layoutBtn.isEnabled = true
                    binding.btnNext.layoutBtn.alpha = 1f
                    binding.btnNext.progressIndicator.visibility = View.GONE
                    val deviceCaps = provisionManager.espDevice.deviceCapabilities
                    if (deviceCaps.contains(AppConstants.CAPABILITY_WIFI_SCAN)) {
                        goToWiFiScanListActivity()
                    } else if (deviceCaps.contains(AppConstants.CAPABILITY_THREAD_SCAN)) {
                        goToThreadScanActivity(true)
                    } else if (deviceCaps.contains(AppConstants.CAPABILITY_THREAD_PROV)) {
                        goToThreadScanActivity(false)
                    } else {
                        goToWiFiConfigActivity()
                    }
                }
            }

            override fun onFailure(e: Exception) {
                e.printStackTrace()
                runOnUiThread {
                    binding.btnNext.layoutBtn.isEnabled = true
                    binding.btnNext.layoutBtn.alpha = 1f
                    binding.btnNext.progressIndicator.visibility = View.GONE
                    binding.tvErrorPop.visibility = View.VISIBLE
                }
            }
        })
    }

    private fun initViews() {
        setToolbar()
        binding.btnNext.textBtn.setText(R.string.btn_next)
        binding.btnNext.layoutBtn.setOnClickListener(nextBtnClickListener)
    }

    private fun setToolbar() {
        setSupportActionBar(binding.titleBar.toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
        supportActionBar!!.setTitle(R.string.title_activity_pop)
        binding.titleBar.toolbar.navigationIcon =
            AppCompatResources.getDrawable(this, R.drawable.ic_arrow_left)
        binding.titleBar.toolbar.setNavigationOnClickListener {
            provisionManager.espDevice?.disconnectDevice()
            finish()
        }
    }

    private fun goToWiFiScanListActivity() {
        val wifiListIntent = Intent(applicationContext, WiFiScanActivity::class.java)
        wifiListIntent.putExtras(intent)
        startActivity(wifiListIntent)
        finish()
    }

    private fun goToThreadScanActivity(scanCapAvailable: Boolean) {
        val threadConfigIntent = Intent(applicationContext, ThreadConfigActivity::class.java)
        threadConfigIntent.putExtras(intent)
        threadConfigIntent.putExtra(AppConstants.KEY_THREAD_SCAN_AVAILABLE, scanCapAvailable)
        startActivity(threadConfigIntent)
        finish()
    }

    private fun goToWiFiConfigActivity() {
        val wifiConfigIntent = Intent(applicationContext, WiFiConfigActivity::class.java)
        wifiConfigIntent.putExtras(intent)
        startActivity(wifiConfigIntent)
        finish()
    }

    private fun showAlertForDeviceDisconnected() {
        val builder = AlertDialog.Builder(this)
        builder.setCancelable(false)
        builder.setTitle(R.string.error_title)
        builder.setMessage(R.string.dialog_msg_ble_device_disconnection)

        // Set up the buttons
        builder.setPositiveButton(
            R.string.btn_ok
        ) { dialog, which ->
            dialog.dismiss()
            finish()
        }
        builder.show()
    }
}
